### PR TITLE
feat: add graph_stores, impl Simple KG & Nebula KG

### PR DESCRIPTION
**Pull request by wey-gu**
draft for RFC #1318

- add graph_stores
- implement Simple kg and NebulaGraph kg

WIP:
- [x] get comments from jerry
- [x] continue to rebase for changes in 0.6.0(introduce graph_store to storage context)
- [x] ~~implement load_from_disk and save_to_disk~~ not needed after 0.6.x

### CODE REVIEW

**Comment by jerryjliu** [comment]
@wey-gu this is cool, thanks for the PR - will try to take some time today to review and offer suggestions on how to rebase

**Comment by Disiok** [comment]
Hey @wey-gu this is great! As @jerryjliu noted, we made some fairly significant changes to how we handle storage (see https://gpt-index.readthedocs.io/en/latest/how_to/storage.html)

**Comment by Disiok** [comment]
Happy to help and answer any questions you might have about 0.6.0

**Comment by jerryjliu** [comment]
awesome yeah @wey-gu adding on to what @Disiok said, it's possible a new graph store abstraction would be mostly used for specific indices, like our knowledge graph index. can be an optional part of our StorageContext that's none by default

**Comment by wey-gu** [comment]
Dear @Disiok 

> Some specific notes:
> 
> * for storage based on external connection (e.g. `NebulaGraph`) here, we no longer try to save the configuration, instead we ask user to reconstruct the connection

wey: Got it, now I don't have to implement that :)

> * we created a `storage context` that bundles docstore, index store, and vector store. The question here is whether if we should add another `graph store` object into it. Would be great if you take a look at 0.6.0 and let us know your thoughts.

wey: This new abstraction is awesome, is it possible to enable chain-able storage context? The knowledge_graph index(now I added graph_store) comes with embedding support, which is memory based, is it possible to enable this embedding storage inside the knowledge_graph index consuming storage context in the future?

Dear @jerryjliu 

> awesome yeah @wey-gu adding on to what @Disiok said, it's possible a new graph store abstraction would be mostly used for specific indices, like our knowledge graph index. can be an optional part of our StorageContext that's none by default

wey: now the graph_store was introduced towards StorageContext and knowledge_graph_index was adapted to be based on simpleGraphStore or nebulaGraphStore :)

-----------

What do you think please of this change?

I will be working on more typical stories/demos for documents to help users understand how it works and how it helps
- better consuming global/cross-node context
- consuming existing knowledge graph in a custom retriever
- composable index co-existing with other indexes
- build a knowledge graph by simply dragging/putting docs in

in separate doc PR + blogs after this is merged.

Thanks again! I am super excited about this change :D

BR//Wey

**Comment by wey-gu** [comment]
Sorry, will fix the lint and UT issue.

**Comment by wey-gu** [comment]
linting and UT passed now.
cc @Disiok @jerryjliu

**Comment by wey-gu** [comment]
pushed another version to address the conflicts related to changes from GPTKnowledgeGraphIndex to KnowledgeGraphIndex by @Disiok

**Comment by wey-gu** [comment]
Thanks @logan-markewich for helping with the review!

Also, another rebase to resolve the lint error..

File: llama_index/graph_stores/simple.py
Comment (logan-markewich) [inline llama_index/graph_stores/simple.py:1]: So basically this file replaces the old `rel_map` in the index struct

Just curious, but I'm wondering how hard it would be to maintain some backwards compatibility here? Some kind of migration script for migrating an only index to a graph store would be very handy, since these indexes can be a little expensive to build

File: llama_index/indices/knowledge_graph/base.py
Comment (logan-markewich) [inline llama_index/indices/knowledge_graph/base.py:128]: nit: Is `table={}` needed? It should hopefully already be defaulting to a dict

File: llama_index/indices/knowledge_graph/retrievers.py
Comment (logan-markewich) [inline llama_index/indices/knowledge_graph/retrievers.py:144]: Just to confirm my understanding, this expands the search of the kg by including the neighboring relations? Or how does this differ from `graph_store_query_depth` ?

File: llama_index/indices/knowledge_graph/retrievers.py
Comment (logan-markewich) [inline llama_index/indices/knowledge_graph/retrievers.py:1]: nit: can we re-name to `retriever.py` ? Since there is only one

File: llama_index/storage/storage_context.py
Comment (logan-markewich) [inline llama_index/storage/storage_context.py:83]: hmmm, this will likely break backwards-compatibility for sure. 

I think this entire section could be more robust, as they should check for the file before loading it.

Right now I get an error like this when loading a previously saved index: `ValueError: No existing llama_index.graph_stores.simple found at ./nodes_index/graph_store.json, skipping load.`

I can take a stab at this or you can as well :)

**Review by logan-markewich — COMMENTED** [review]
This looks super good, and thanks for tackling such a complicated PR!

A few minor nits, but my main worry is backwards compatibility (both with previously saved knowledge graph indexes, as well as previously saved indexes in general)

The storage context code likely needs a bit of TLC to be more robust

File: examples/composable_indices/city_analysis/PineconeDemo-CityAnalysis.ipynb
Comment (wey-gu) [reply examples/composable_indices/city_analysis/PineconeDemo-CityAnalysis.ipynb:1]: Sorry for the silly mistake, will remove

File: llama_index/__init__.py
Comment (wey-gu) [reply llama_index/__init__.py:28]: Sure, will do so for all with KnowledgeGraphIndex

File: llama_index/data_structs/data_structs.py
Comment (wey-gu) [reply llama_index/data_structs/data_structs.py:210]: Sure, sorry and thanks!

File: llama_index/data_structs/data_structs.py
Comment (wey-gu) [reply llama_index/data_structs/data_structs.py:213]: sure, thanks

File: llama_index/data_structs/data_structs.py
Comment (wey-gu) [reply llama_index/data_structs/data_structs.py:235]: sure, thanks

File: llama_index/graph_stores/simple.py
Comment (wey-gu) [reply llama_index/graph_stores/simple.py:26]: Thanks! will fix

File: llama_index/graph_stores/types.py
Comment (wey-gu) [reply llama_index/graph_stores/types.py:10]: Sure, will do so, thanks!

File: llama_index/indices/knowledge_graph/retrievers.py
Comment (wey-gu) [reply llama_index/indices/knowledge_graph/retrievers.py:1]: Yes, will do so :)

File: llama_index/indices/knowledge_graph/retrievers.py
Comment (wey-gu) [reply llama_index/indices/knowledge_graph/retrievers.py:144]: Great question Logan, they are somewhat similar indeed, while this option enables more knowledge to be involved as it adds node-level locality: the kg(with n depth) traverse from subj not just question keywords(say x), but also from subjs in nodes with question-keywords, i.e. node 96 with x, so `x -> k` , `a -> d` can be added.

Here `a -> d` is actually already added when depth >= 2, but the knowledge of `x -> k` is only added when this option enabled.

llama_index_global_explore


```
┌──────────────────┬──────────────────┬──────────────────┬──────────────────┐
│ .─.       .─.    │  .─.       .─.   │            .─.   │  .─.       .─.   │
│( x )─────▶ y )   │ ( x )─────▶ a )  │           ( j )  │ ( m )◀────( x )  │
│ `▲'       `─'    │  `─'       `─'   │            `─'   │  `─'       `─'   │
│  │     1         │        2         │        3    │    │        4         │
│ .─.              │                  │            .▼.   │                  │
│( z )─────────────┼──────────────────┼──────────▶( i )─┐│                  │
│ `◀────┐          │                  │            `─'  ││                  │
├───────┼──────────┴──────────────────┴─────────────────┼┴──────────────────┤
│       │                      Docs/Knowledge           │                   │
│       │                            ...                │                   │
│       │                                               │                   │
├───────┼──────────┬──────────────────┬─────────────────┼┬──────────────────┤
│  .─.  └──────.   │  .─.        .─.  │                 ││  .─.             │
│ ( x )─────( b )  │ ( a )      ( x ) │                 └┼▶( n )            │
│  `─'       `─'   │  `─'        `─'  │                  │  `─'             │
│        95   │    │   │    96    │   │                  │   │              │
│            .▼.   │  .▼.        .▼.  │                  │   ▼              │
│           ( c )  │ ( d )      ( k ) │                  │  .─.             │
│            `─'   │  `─'        `─'  │                  │ ( x )            │
└──────────────────┴──────────────────┴──────────────────┴──`─'─────────────┘
```

Do you think `explore_global_knowledge` is properly-named and worthy to be added here?

Thanks!

File: llama_index/indices/knowledge_graph/retrievers.py
Comment (wey-gu) [reply llama_index/indices/knowledge_graph/retrievers.py:245]: Sure, it could be, we need to revisit this and at least introduce more mechanisms to refine, deduplicate and truncate, with your reminding, I will add in this round of push and will create e2e demo apps with it after it is merged, then I/we will come with more ideas/experiences to polish here, too.

File: llama_index/graph_stores/simple.py
Comment (wey-gu) [reply llama_index/graph_stores/simple.py:1]: Indeed, make a lot of sense, I didn't think of this backward compatibility yet, but I'll try so in this round of push. thanks!

**Review by wey-gu — COMMENTED** [review]
Will work on all great review/improvement points from Logan, thanks again!

File: llama_index/graph_stores/simple.py
Comment (wey-gu) [reply llama_index/graph_stores/simple.py:92]: Thanks, agreed!

File: llama_index/indices/knowledge_graph/base.py
Comment (wey-gu) [reply llama_index/indices/knowledge_graph/base.py:128]: True, KG already handles it properly, it's removed now.

File: llama_index/storage/storage_context.py
Comment (wey-gu) [reply llama_index/storage/storage_context.py:83]: Dear @logan-markewich, I pushed a commit with all comments covered except for this part, I'll go for lunch first and come back soon for this part.

**Comment by wey-gu** [comment]
> This looks super good, and thanks for tackling such a complicated PR!
> 
> A few minor nits, but my main worry is backwards compatibility (both with previously saved knowledge graph indexes, as well as previously saved indexes in general)
> 
> The storage context code likely needs a bit of TLC to be more robust
> 
> Lastly, and maybe not this PR, but some docs on using Nebula would be cool. (I actually still need to try setting it up lol)

Dear @logan-markewich ,

- For backward compatibility, I am trying to add previously saved knowledge graph indexes support(or provide a converting tool) today.
- I will revisit and improve the storage context part now
- I'll add docs of NebulaGraph later in upcoming PRs! For now, in case you are about/interested in setting it up, on your own machine, you could go with [docker extension](https://hub.docker.com/extensions/weygu/nebulagraph-dd-ext), and on a server with docker, you could go with [nebula-up](https://github.com/wey-gu/nebula-up), both of which are one-liner or one-click, and then we could create needed space and schema from NebulaGraph console or Studio following existing notebook's "Prepare for NebulaGraph" chapter.

File: llama_index/indices/knowledge_graph/retrievers.py
Comment (wey-gu) [reply llama_index/indices/knowledge_graph/retrievers.py:146]: In the future, I am thinking of adding vector index/search in NebulaGraph, in that case, the node(doc chunk) could be persisted as another type of vertex in NebulaGraph(with its embeddings) and we could do topk distance search on the graph between vertices(for both subj, obj and doc chunks), in that case, if we wire the chunks--> subjs relationships in NebulaGrpah, then the wo could somewhat be almost the same/much similar as your understanding from intuition.

I am still studying this(all-in-one graph+doc+vector storage) now, what do you think, please?

File: llama_index/indices/knowledge_graph/retrievers.py
Comment (logan-markewich) [reply llama_index/indices/knowledge_graph/retrievers.py:258]: Makes sense!

**Comment by logan-markewich** [comment]
I think this is good to ship! Looking forward to your future work @wey-gu ! Knowledge graphs can be very powerful, and hoping llama-index can continue to be a great tool to leverage them

**Review by logan-markewich — APPROVED** [review]

File: llama_index/data_structs/struct_type.py
Comment (Disiok) [inline llama_index/data_structs/struct_type.py:87]: nit: where are these used?

File: llama_index/indices/base.py
Comment (Disiok) [inline llama_index/indices/base.py:63]: I don't think the base class should know about graph store?

File: llama_index/indices/knowledge_graph/base.py
Comment (Disiok) [inline llama_index/indices/knowledge_graph/base.py:61]: nit: if graph store is already a part of the storage context, should we still accept it separately here?

File: llama_index/indices/knowledge_graph/base.py
Comment (Disiok) [inline llama_index/indices/knowledge_graph/base.py:109]: I don't think we need to inject into kwargs. We already have the reference to the index, which exposes the graph store?

File: llama_index/indices/knowledge_graph/retriever.py
Comment (Disiok) [inline llama_index/indices/knowledge_graph/retriever.py:79]: nit: why don't we use the `index.graph_store` reference?

File: llama_index/storage/storage_context.py
Comment (Disiok) [inline llama_index/storage/storage_context.py:72]: just checking: there should be minimal overhead in doing this right?

Otherwise, we should just not construct an object and leave it as None, so it doesn't impact other type of indices that don't use graph.

**Review by Disiok — APPROVED** [review]
Added some nitpick comments, but I don't want to block merging this massive PR.
We can do more cleanup after landing as well.

File: llama_index/storage/storage_context.py
Comment (logan-markewich) [reply llama_index/storage/storage_context.py:72]: It's the same as a vector_store essentially in terms of impact -> should be near-zero. 

Similar to how other indexes don't use the vector store, but we still instantiate it I guess

File: llama_index/indices/base.py
Comment (logan-markewich) [reply llama_index/indices/base.py:63]: mmm true, but could say the same about the vector store? We can take a look at both in a future PR

File: llama_index/data_structs/struct_type.py
Comment (logan-markewich) [reply llama_index/data_structs/struct_type.py:87]: I think this just follows the same structure for the vector store registry -- is that registry used anywhere?

**Comment by logan-markewich** [comment]
Addressed a majority of your comments, I think it's good to land for now :)

**Comment by jerryjliu** [comment]
@wey-gu this is an amazing change, thanks for the contribution. thanks to @logan-markewich @Disiok for the reviews too. just a heads up, planning to publish this friday morning pacific time - let me know your twitter handle!

**Comment by wey-gu** [comment]
> @wey-gu this is an amazing change, thanks for the contribution. thanks to @logan-markewich @Disiok for the reviews too. just a heads up, planning to publish this friday morning pacific time - let me know your twitter handle!

Dear @jerryjliu ,
Thanks so much, I am honored to have the chance to bring something the Llama Index project and it's been an awesome experience working within the great Llama community with you, @logan-markewich and @Disiok , I am working on upcoming PR and demo-project/video on top of this change.

My handle is wey_gu :)

Thanks!

**Comment by wey-gu** [comment]
> I think this is good to ship! Looking forward to your future work @wey-gu ! Knowledge graphs can be very powerful, and hoping llama-index can continue to be a great tool to leverage them

Dear @logan-markewich ,

Many thanks for the great help and guide(and big thanks to @Disiok !!)!! I am preparing for upcoming PRs/DEMOs, let's make LLMs understand more knowledge with graphs!

BR//Wey

File: llama_index/graph_stores/nebulagraph.py
Comment (wey-gu) [reply llama_index/graph_stores/nebulagraph.py:282]: Could you please upgrade to NebulaGraph 3.5.0, and see what happens, this implementation was expecting NebulaGraph 3.5.0+

**Comment by pachgadehardik** [comment]
Need to load the existing data indexes from nebula graph into the retriever for knowledge graph query_engine. Is that functionality available in llama-index ?

**Comment by wey-gu** [comment]
> Need to load the existing data indexes from nebula graph into the retriever for knowledge graph query_engine. Is that functionality available in llama-index ?

For text2cypher, yes, it's already implemented! Follow https://gpt-index.readthedocs.io/en/latest/examples/query_engine/knowledge_graph_query_engine.html is all needed.

For Graph RAG(find major entities with keyword or embedding from the task, get subgraph as context), not yet fully supported. I am thinking of adding this soon.

**Comment by pachgadehardik** [comment]
> > Need to load the existing data indexes from nebula graph into the retriever for knowledge graph query_engine. Is that functionality available in llama-index ?
> 
> @pachgadehardik 
> 
> For text2cypher, yes, it's already implemented! Follow https://gpt-index.readthedocs.io/en/latest/examples/query_engine/knowledge_graph_query_engine.html is all needed.
> 
> For Graph RAG(find major entities with keyword or embedding from the task, get subgraph as context), not yet fully supported. I am thinking of adding this soon.

Thanks a lot @wey-gu  for the update. However  I am facing an issue while running KnowledgeGraphQueryEngine. NebulaGraphStore is being loaded but when executing the KGQueryEngine, was facing a query syntax - 
ValueError: Query failed. Query: 
MATCH ()-[e:`relationship`]->()
  WITH e limit 1
MATCH (m)-[:`relationship`]->(n) WHERE id(m) == src(e) AND id(n) == dst(e)
RETURN "(:" + tags(m)[0] + ")-[:relationship]->(:" + tags(n)[0] + ")" AS rels
, Param: {}Error message: Scan vertices or edges need to specify a limit number, or limit number cannot push down.

**Comment by wey-gu** [comment]
> > > Need to load the existing data indexes from nebula graph into the retriever for knowledge graph query_engine. Is that functionality available in llama-index ?
> > 
> > 
> > @pachgadehardik
> > For text2cypher, yes, it's already implemented! Follow https://gpt-index.readthedocs.io/en/latest/examples/query_engine/knowledge_graph_query_engine.html is all needed.
> > For Graph RAG(find major entities with keyword or embedding from the task, get subgraph as context), not yet fully supported. I am thinking of adding this soon.
> 
> Thanks a lot @wey-gu for the update. However I am facing an issue while running KnowledgeGraphQueryEngine. NebulaGraphStore is being loaded but when executing the KGQueryEngine, was facing a query syntax - ValueError: Query failed. Query: MATCH ()-[e:`relationship`]->() WITH e limit 1 MATCH (m)-[:`relationship`]->(n) WHERE id(m) == src(e) AND id(n) == dst(e) RETURN "(:" + tags(m)[0] + ")-[:relationship]->(:" + tags(n)[0] + ")" AS rels , Param: {}Error message: Scan vertices or edges need to specify a limit number, or limit number cannot push down.

Dear @pachgadehardik 

Could you share the NebulaGraph version? If it's an older version of NebulaGraph like 3.1.0, it's highly recommended to be upgraded to NebulaGraph 3.5.0, this is basically just a binary replacement(offline)

**Comment by pachgadehardik** [comment]
> > > > Need to load the existing data indexes from nebula graph into the retriever for knowledge graph query_engine. Is that functionality available in llama-index ?
> > > 
> > > 
> > > @pachgadehardik
> > > For text2cypher, yes, it's already implemented! Follow https://gpt-index.readthedocs.io/en/latest/examples/query_engine/knowledge_graph_query_engine.html is all needed.
> > > For Graph RAG(find major entities with keyword or embedding from the task, get subgraph as context), not yet fully supported. I am thinking of adding this soon.
> > 
> > Thanks a lot @wey-gu for the update. However I am facing an issue while running KnowledgeGraphQueryEngine. NebulaGraphStore is being loaded but when executing the KGQueryEngine, was facing a query syntax - ValueError: Query failed. Query: MATCH ()-[e:`relationship`]->() WITH e limit 1 MATCH (m)-[:`relationship`]->(n) WHERE id(m) == src(e) AND id(n) == dst(e) RETURN "(:" + tags(m)[0] + ")-[:relationship]->(:" + tags(n)[0] + ")" AS rels , Param: {}Error message: Scan vertices or edges need to specify a limit number, or limit number cannot push down.
> 
> Dear @pachgadehardik 
> 
> Could you share the NebulaGraph version? If it's an older version of NebulaGraph like 3.1.0, it's highly recommended to be upgraded to NebulaGraph 3.5.0, this is basically just a binary replacement(offline)

@wey-gu, the current version deployed in kubernetes cluster is 3.4.0

**Comment by wey-gu** [comment]
> > > > > Need to load the existing data indexes from nebula graph into the retriever for knowledge graph query_engine. Is that functionality available in llama-index ?
> > > > 
> > > > 
> > > > @pachgadehardik
> > > > For text2cypher, yes, it's already implemented! Follow https://gpt-index.readthedocs.io/en/latest/examples/query_engine/knowledge_graph_query_engine.html is all needed.
> > > > For Graph RAG(find major entities with keyword or embedding from the task, get subgraph as context), not yet fully supported. I am thinking of adding this soon.
> > > 
> > > 
> > > Thanks a lot @wey-gu for the update. However I am facing an issue while running KnowledgeGraphQueryEngine. NebulaGraphStore is being loaded but when executing the KGQueryEngine, was facing a query syntax - ValueError: Query failed. Query: MATCH ()-[e:`relationship`]->() WITH e limit 1 MATCH (m)-[:`relationship`]->(n) WHERE id(m) == src(e) AND id(n) == dst(e) RETURN "(:" + tags(m)[0] + ")-[:relationship]->(:" + tags(n)[0] + ")" AS rels , Param: {}Error message: Scan vertices or edges need to specify a limit number, or limit number cannot push down.
> > 
> > 
> > Dear @pachgadehardik
> > Could you share the NebulaGraph version? If it's an older version of NebulaGraph like 3.1.0, it's highly recommended to be upgraded to NebulaGraph 3.5.0, this is basically just a binary replacement(offline)
> 
> @wey-gu, the current version deployed in kubernetes cluster is 3.4.0

I see, the current schema fetching way is not compatible with the cluster that's older than 3.5.0, I'll pr to fix this today!
Thanks for letting me know this and sorry for it!

**Comment by wey-gu** [comment]
> pachgadehardik

It should be fixed in #7204

**Pull request merged**
